tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/OptimizerCore.scala

```diff
@@ -3690,14 +3683,21 @@ private[optimizer] object OptimizerCore {
   private def getIntrinsicCode(target: AbstractMethodID): Int =
     Intrinsics.intrinsics(target.toString)
 
-  private trait State[A] {
-    def makeBackup(): A
-    def restore(backup: A): Unit
+  private trait StateBackup {
+    def restore(): Unit
   }
 
-  private class SimpleState[A](var value: A) extends State[A] {
-    def makeBackup(): A = value
-    def restore(backup: A): Unit = value = backup
+  private trait State {
+    def makeBackup(): StateBackup
+  }
+
+  private class SimpleState[A](var value: A) extends State {
+    def makeBackup(): StateBackup = {
+      val savedValue = value
+      new StateBackup {
+        override def restore(): Unit = value = savedValue
+      }
```

Comment:
Consider using a private inner class with a parameter. It might be easier to understand how the capturing works and is even a line less:

```scala
private class Backup(savedValue: A) extends StateBackup {
  override def restore(): Unit = value = savedValue
}

def makeBackup(): StateBackup = new Backup(value)
```